src/apps/relay/netengine.c

```diff
@@ -1816,57 +2100,67 @@ void setup_server(void) {
+    setup_barriers();
 
-  if (turn_params.net_engine_version == NEV_UDP_SOCKET_PER_THREAD)
-    setup_socket_per_thread_udp_listener_servers();
-  else if (turn_params.net_engine_version == NEV_UDP_SOCKET_PER_ENDPOINT)
-    setup_socket_per_endpoint_udp_listener_servers();
-  else if (turn_params.net_engine_version == NEV_UDP_SOCKET_PER_SESSION)
-    setup_socket_per_session_udp_listener_servers();
+    nRet = setup_general_relay_servers();
+    if (nRet)
+      break;
 
-  if (turn_params.net_engine_version != NEV_UDP_SOCKET_PER_THREAD) {
-    setup_tcp_listener_servers(turn_params.listener.ioa_eng, NULL);
-  }
+    if (turn_params.net_engine_version == NEV_UDP_SOCKET_PER_THREAD)
+      setup_socket_per_thread_udp_listener_servers();
+    else if (turn_params.net_engine_version == NEV_UDP_SOCKET_PER_ENDPOINT) {
+      nRet = setup_socket_per_endpoint_udp_listener_servers();
+      if (nRet)
+        break;
+    } else if (turn_params.net_engine_version == NEV_UDP_SOCKET_PER_SESSION)
+      setup_socket_per_session_udp_listener_servers();
 
-  {
-    int tot = 0;
-    if (udp_relay_servers[0]) {
-      tot = get_real_udp_relay_servers_number();
-    }
-    if (tot) {
-      TURN_LOG_FUNC(TURN_LOG_LEVEL_INFO, "Total UDP servers: %d\n", (int)tot);
+    if (turn_params.net_engine_version != NEV_UDP_SOCKET_PER_THREAD) {
+      setup_tcp_listener_servers(turn_params.listener.ioa_eng, NULL);
     }
-  }
 
-  {
-    int tot = get_real_general_relay_servers_number();
-    if (tot) {
-      int i;
-      for (i = 0; i < tot; i++) {
-        if (!(general_relay_servers[i])) {
-          TURN_LOG_FUNC(TURN_LOG_LEVEL_ERROR, "General server %d is not initialized !\n", (int)i);
+    if (udp_relay_servers[0])
+      TURN_LOG_FUNC(TURN_LOG_LEVEL_INFO, "Total turn servers(UDP): %d\n", (int)get_real_udp_relay_servers_number());
+
+    TURN_LOG_FUNC(TURN_LOG_LEVEL_INFO, "Total turn servers(TCP): %d\n", (int)get_real_general_relay_servers_number());
+    {
+      int tot = get_real_general_relay_servers_number();
+      if (tot) {
+        int i;
+        for (i = 0; i < tot; i++) {
+          if (!(general_relay_servers[i])) {
+            TURN_LOG_FUNC(TURN_LOG_LEVEL_ERROR, "General turn server(TCP) %d is not initialized !\n", (int)i);
+          }
         }
       }
     }
-  }
 
-  {
-    authserver_id sn = 0;
-    for (sn = 0; sn < authserver_number; ++sn) {
-      authserver[sn].id = sn;
-      setup_auth_server(&(authserver[sn]));
+    {
+      authserver_id sn = 0;
+      for (sn = 0; sn < authserver_number; ++sn) {
+        authserver[sn].id = sn;
+        setup_auth_server(&(authserver[sn]));
+      }
+      TURN_LOG_FUNC(TURN_LOG_LEVEL_INFO, "Total auth threads: %d\n", authserver_number);
     }
-    TURN_LOG_FUNC(TURN_LOG_LEVEL_INFO, "Total auth threads: %d\n", authserver_number);
-  }
 
-  setup_admin_server();
+    nRet = setup_admin_server();
+    if (nRet)
+      break;
 
-  barrier_wait();
+    barrier_wait();
+    return 0;
+
+  } while (0);
```

Comment:
this is a bizare construction. and not really what do{}while(); is for.

Why not use a goto instead?